lua/impatient/cachepack.lua

```diff
@@ -0,0 +1,90 @@
+local ffi = require('ffi')
+
+local M = {}
+
+-- using double for packing/unpacking numbers has no conversion overhead
+local c_double = ffi.typeof("double[1]")
+local sizeof_c_double = ffi.sizeof("double")
+
+local OutputBuffer = {}
+
+function OutputBuffer.create() return {} end
+
+function OutputBuffer.write_number(buf, num)
+  buf[#buf+1] = ffi.string(c_double(num), sizeof_c_double)
+end
+
+function OutputBuffer.write_string(buf, str)
+  OutputBuffer.write_number(buf, #str)
+  buf[#buf+1] = str
+end
+
+function OutputBuffer.to_string(buf)
+  return table.concat(buf)
+end
+
+local InputBuffer = {}
+
+function InputBuffer.create(str)
+  return {
+    ptr = ffi.new("const char[?]", #str, str),
+    pos = 0,
+    size = #str
```

Comment:
You may want to cache the value of `#str` so it isn't calculated twice.

Reply:
Lua stores the size of a string as far as I'm aware, unlike C which is null-terminator based, so I think this isn't strictly necessary, but you can certainly change things to your liking.